### PR TITLE
serial sync from NMEA devices (GPS, ES8266/nodemcu with NTP)

### PR DESCRIPTION
**Pull request by venustrg**
just current state of nmea sync branch.
es8266/init.lua generates valid NMEA with crc every 8 hours.
stc15w408 can parse NMEA, check validity & crc, set time including DOW.
setting time paused at least for 4 hrs.
used time.h structures but there is no enough mem for time->unix, applying TZ, unix->time.
tried others time functions like avr libc or https://www.oryx-embedded.com/doc/date__time_8c_source.html without success, still no enough mem.

docs were not updated after ntp_sync with proprietary serial protocol. just current state.

### CODE REVIEW

**Comment by venustrg** [comment]
standard sdcc time.h functions mktime() and localtime() with TZ correction just commented out.
right now clock works fine, nodemcu makes TZ correction and sends NMEA with local time.

File: Makefile [hunk 1]
```diff
@@ -1,14 +1,15 @@
 SDCC ?= sdcc
-STCCODESIZE ?= 4089
-SDCCOPTS ?= --code-size $(STCCODESIZE) --xram-size 0 --data-loc 0x30 --disable-warning 126 --disable-warning 59
-SDCCREV ?= -Dstc15f204ea
+STCCODESIZE ?= 8184
```
Comment (zerog2k) [inline Makefile:2]: currently we have about three different mcu variants:
1. stc15f204ea (4k code, 256b ram, no xram)
2. stc15w404as (4k code, 256b ram, 256b xram)
3. stc15w408as (8k code, 256b ram, 256b xram).
So while the stack/ram size opts are generally ok for 4xx series, they wont work properly on 2xx series.

Additionally complicating this, is that while there is this define in the code, e.g. "stc15w408as", it was unfortunately used to mean both the mcu and some options of certain clock hardware options (e.g. 3rd button). What I really need to do is to separate out the mcu options from the hardware variant options (in hwconfig.h) as the mcus are generally interchangeable, while the hardware options are not.

File: Makefile [hunk 2]
```diff
@@ -1,14 +1,15 @@
...
+SDCCOPTS ?= --code-size $(STCCODESIZE) --stack-auto --iram-size 256 --xram-size 256 --data-loc 0x30 --disable-warning 126 --disable-warning 59
+SDCCREV ?= -Dstc15w408as
 STCGAL ?= stcgal/stcgal.py
 STCGALOPTS ?= 
 STCGALPORT ?= /dev/ttyUSB0
 STCGALPROT ?= auto
 FLASHFILE ?= main.hex
 SYSCLK ?= 11059
-CFLAGS ?= -DWITH_ALT_LED9 -DWITHOUT_LEDTABLE_RELOC -DSHOW_TEMP_DATE_WEEKDAY
+CFLAGS ?= -DWITH_ALT_LED9 -DWITHOUT_LEDTABLE_RELOC -DSHOW_TEMP_DATE_WEEKDAY -DWITH_CAPACITOR -DWITH_NMEA
```
Comment (zerog2k) [inline Makefile:11]: I think `WITH_CAPACITOR` will ask DS1302 to trickle charge the battery, yes? I don't think that should be default, as most clocks have lithium primary cell which should not receive any charge current.

**Comment by venustrg** [comment]
i do not propose making the capacitor and 4xx options by default. i just left the makefile as i used it to compile, so you can see under what conditions I don’t have enough memory to implement tz handling.

i understand this is not final version of syncing code

**Comment by venustrg** [comment]
by the way, as far as i know, original chinese fw enables trickle charging by default and this kills the primary cells quickly. :)

**Comment by zerog2k** [comment]
>i do not propose making the capacitor and 4xx options by default. i just left the makefile as i used it to compile, so you can see under what conditions I don’t have enough memory to implement tz handling.
Ok thanks I understand now. I'll wait til you are ready to merge it.
Thanks again for this great contribution!

**Comment by venustrg** [comment]
now we have TZ/DST options, stored to EEPROM, loaded at startup.
still have >1k to implement TZ handling without converting to unixtime.

Name             Start    End      Size     Max     
---------------- -------- -------- -------- --------
ROM/EPROM/FLASH  0x0000   0x1ad0    6865     8184

**Comment by venustrg** [comment]
wrote from scratch tz processing for struct tm and looks like it works correctly.
supported tz are +6:00, -4:30, +5:45 etc.
negative values indicated with right point (point3).
eeprom will be reset after flashing, so tz will be -1 by default.
```
   Name             Start    End      Size     Max     
   ---------------- -------- -------- -------- --------
   ROM/EPROM/FLASH  0x0000   0x1fd3    8148     8184   
```

**Comment by venustrg** [comment]
Makefile reverted to default

**Comment by venustrg** [comment]
looks like sync is working fine.
i have no more complaints.
please check.

**Comment by venustrg** [comment]
one sync attempt per hour looks reasonable.
ESP8266 sleeps between tries when using init-cool.lua and doesn't heat up. 
clock fw limits sync to 1 per 6 hours (MIN_NMEA_PAUSE in nmea.h).
it seems to work flawlessly.

**Comment by venustrg** [comment]
i have nothing more to commit.
the patch looks finished. :)

it looks like I burned my DS1302 out due to an incorrect UART-TTL connection. now it doesn't see the capacitor and resets every time the power is turned off. i'm waiting for a new ds1302 with aliexpress.

**Comment by venustrg** [comment]
Have you tried this patch with GPS / NTP?
my watch seems to be working stably.
I also ordered the NEO-6M GPS module for testing, but haven’t received it yet.

**Comment by zerog2k** [comment]
> Have you tried this patch with GPS / NTP?
> my watch seems to be working stably.
> I also ordered the NEO-6M GPS module for testing, but haven’t received it yet.

Hi George, no sorry, I haven't had a chance to get to this due to home and work priorities ;)
I'm hoping to get some free time to test this weekend.

**Comment by zerog2k** [comment]
I got a chance to test this out on my clock, and while it works, it never seems to sync to my gps.
Here is example output from my gps with location redacted.

```
$SkyTraq,Venus8
$Kernel,v2.2.0,00003C92,1626,00008005,I,16.367668MHz
$ver,1.8.27,rev,15.05.21

$GPGGA,120005.000,0000.0000,N,00000.0000,E,0,00,0.0,0.0,M,0.0,M,,0000*6B
$GPGLL,0000.0000,N,00000.0000,E,120005.000,V,N*47
$GPGSA,A,1,,,,,,,,,,,,,0.0,0.0,0.0*30
$GPRMC,120005.000,V,0000.0000,N,00000.0000,E,000.0,000.0,280606,,,N*7A
$GPVTG,000.0,T,,M,000.0,N,000.0,K,N*02
$GPZDA,120005.000,28,06,2006,00,00*58

... after one second ...

$GPGGA,120006.000,0000.0000,N,00000.0000,E,0,00,0.0,0.0,M,0.0,M,,0000*68
$GPGLL,0000.0000,N,00000.0000,E,120006.000,V,N*44
$GPGSA,A,1,,,,,,,,,,,,,0.0,0.0,0.0*30
$GPRMC,120006.000,V,0000.0000,N,00000.0000,E,000.0,000.0,280606,,,N*79
$GPVTG,000.0,T,,M,000.0,N,000.0,K,N*02
$GPZDA,120006.000,28,06,2006,00,00*5B

... repeats ...

... after several minutes ...

$GPGGA,224245.000,xx15.1813,N,yyy30.5860,W,1,06,4.0,200.9,M,-27.7,M,,0000*67
$GPGLL,xx15.1813,N,yyy30.5860,W,224245.000,A,A*44
$GPGSA,A,3,06,17,05,28,19,02,,,,,,,7.4,4.0,6.2*34
$GPRMC,224245.000,A,xx15.1813,N,yyy30.5860,W,000.0,123.1,030520,,,A*76
$GPVTG,123.1,T,,M,000.0,N,000.0,K,A*0C
$GPZDA,224245.000,03,05,2020,00,00*57

... after one second ...

$GPGGA,224246.000,xx15.1813,N,yyy30.5860,W,1,06,4.0,200.9,M,-27.7,M,,0000*64
$GPGLL,xx15.1813,N,yyy30.5860,W,224246.000,A,A*47
$GPGSA,A,3,06,17,05,28,19,02,,,,,,,7.4,4.0,6.2*34
$GPGSV,3,1,10,02,72,310,18,06,56,029,25,12,52,312,14,19,39,072,24*7F
$GPGSV,3,2,10,05,32,179,23,17,23,086,28,25,21,318,15,24,18,243,07*77
$GPGSV,3,3,10,28,06,145,20,09,01,082,*74
$GPRMC,224246.000,A,xx15.1813,N,yyy30.5860,W,000.0,123.1,030520,,,A*75
$GPVTG,123.1,T,,M,000.0,N,000.0,K,A*0C
$GPZDA,224246.000,03,05,2020,00,00*54
$GPGGA,224247.000,xx15.1813,N,yyy30.5860,W,1,06,4.0,200.9,M,-27.7,M,,0000*65

... repeats ...
```

**Comment by venustrg** [comment]
protocol looks correct and should be accepted by watch firmware. i will try to send your lines from my ESP8266.
do you have hex dump of this data? do lines end with "\r\n" or just with "\n" ?  i had no chance to examine data from real gps device,

**Comment by venustrg** [comment]
just added string from your messafe to my init.lua as last time in UpdateRtc()
    uarttimestring = "$GPRMC,224245.000,A,xx15.1813,N,yyy30.5860,W,000.0,123.1,030520,,,A*76\r\n"

and clock synced successfully.

**Comment by venustrg** [comment]
also, what is your serial port speed?
i use 9600 but https://en.wikipedia.org/wiki/NMEA_0183 tell about 4800.
i'm not sure if i should switch to 4800..

**Comment by zerog2k** [comment]
The port speed on this one is 9600.
So I hooked directly up from serial port to P3_6 using uart adapter, and I can reproduce this.

1. If I send the `$GPRMC` as the last sentence, the time will set
2. If there are any other sentences sent _after_ `$GPRMC` it will not work
I am using CR (\r) + LF (\n) line termination.

If I get more time, I'll try to look into why.

**Comment by venustrg** [comment]
i understand now. looks line "REN=0" doesn't stop uart from symbols receiving. replaced with state checking.

**Comment by venustrg** [comment]
can you please check one more time?

**Comment by venustrg** [comment]
i'm checking by myself. i will say if it works

**Comment by venustrg** [comment]
seems to be working now

**Comment by venustrg** [comment]
try when you have time for this

**Comment by zerog2k** [comment]
Ok I managed to get some time to build and load the latest onto my clock. Good news is I see that it tries to set something whenever I dump a batch of nmea sentences to the clock (simulating gps output). However, it does not seem to set date properly. I get strange dates like 25/25 2025 or 11.11 2011. 
I'll still have to do a few more tests (also will leave gps hooked up for a while - takes about 10+m sometimes for it to get a cold lock where GPRMC has "valid" (`A`) status.

**Comment by venustrg** [comment]
i plugged my USB-TTL adapter instead of ESP8266 and used "picocom /dev/ttyUSB0 -b 9600 -c --omap crcrlf" to send the data. tried to copy-paste all NMEA messages from this topic at the same time. it always finds first correct line and synchronizes to 224245/GMT correctly. i can't imagine how the wrong time could be set, it only can be set from correct GPRMC line with CRC checking.

**Comment by venustrg** [comment]
at least i found that my ds 302 was alive, I just broke off the soldering of the ionistor, probably when i connected the usb-ttl. now i have no idea what to do with ordered 10pcs of ds1302. :)

**Comment by zerog2k** [comment]
@venustrg quick question: what version of sdcc are you using?

**Comment by venustrg** [comment]
venus@venus$ sdcc -v
SDCC : mcs51/z80/z180/r2k/r3ka/gbz80/tlcs90/ez80_z80/ds390/pic16/pic14/TININative/ds400/hc08/s08/stm8/pdk13/pdk14/pdk15 4.0.0 #11528 (Linux)

**Comment by zerog2k** [comment]
I asked because I had found that after about 3.6.x something, several 8051 regressions popped up , including some which cause incorrect code (non-atomic SFR access, etc), and I'm using platformio which I think uses a recent 3.9.x something?
Anyhow, I doubt that's the issue I'm seeing, but to rule this out, can you share/attach the .hex file you are using here - I'll try this.

**Comment by venustrg** [comment]
i use WITH_CAPACITOR, attached file was compiled without it
```
STCCODESIZE ?= 8184
SDCCOPTS ?= --code-size $(STCCODESIZE) --stack-auto --iram-size 256 --xram-size 256 --data-loc 0x30 --disable-warning 126 --disable-warning 59
SDCCREV ?= -Dstc15w404as
CFLAGS ?= -DWITH_ALT_LED9 -DWITHOUT_LEDTABLE_RELOC -DSHOW_TEMP_DATE_WEEKDAY -DWITH_NMEA
```
[main.hex.gz](https://github.com/zerog2k/stc_diyclock/files/4618274/main.hex.gz)

**Comment by venustrg** [comment]
downgraded my sdcc to 3.8.0, clock syncs correctly

**Comment by venustrg** [comment]
tested with NEO-6MV2 GPS module, works ok.

gps_testing

**Comment by zerog2k** [comment]
Thanks George for testing with GPS module. I'll continue to work on my local device.

**Pull request merged**